tests/abi/abi.c

```diff
@@ -0,0 +1,175 @@
+    Generated[Counter] = getNextLFSR();
+}
+
+void dump(REGISTER_C_TYPE *Generated) {
+  /* Dump registers */
+  puts("  Registers:");
+#define DUMP_REGISTER(REGISTER)                           \
+  do {                                                    \
+    printf("    - Name: \"%s\"\n      Value: 0x%016lx\n", \
+           STRINGIFY(REGISTER),                           \
+           Generated[RegisterIndex--]);                   \
+  } while (0);
+  {
+    uint32_t RegisterIndex = REGISTER_COUNT - 1;
+    FOR_EACH(DUMP_REGISTER, REGISTER_LIST);
+  }
+#undef DUMP_REGISTER
+
+  /* Dump Stack */
+  puts("  Stack:");
+  for (uint32_t Index = REGISTER_COUNT; Index < GENERATED_COUNT; ++Index)
+    printf("    - Offset: -0x%04x\n      Value: 0x%016lx\n",
+           (GENERATED_COUNT - Index) * REGISTER_SIZE,
+           Generated[Index]);
+}
+
+#define CALL_ITERATION_WITH(FUNCTION_NAME) \
+  CONCATENATE(iterationImplOn_, FUNCTION_NAME)
+
+#define GENERATED_BYTES ((GENERATED_COUNT) * sizeof(REGISTER_C_TYPE))
+
+/* clang-format off */
+#define DEFINE_ITERATION_WITH(FUNCTION_NAME)                                 \
+void CALL_ITERATION_WITH(FUNCTION_NAME)(void) {                              \
+  REGISTER_C_TYPE *Generated = (REGISTER_C_TYPE *) malloc(GENERATED_BYTES);  \
+  generate(Generated);                                                       \
+  dump(Generated);                                                           \
+                                                                             \
+  __asm__ __volatile__ (                                                     \
+    /* Save all the registers */                                             \
+    FOR_EACH(PUSH_REGISTER, REGISTER_LIST)                                   \
+    /* Popularize the stack with random data */                              \
+    REPEAT(PUSH_CONSTANT, STACK_VALUE_COUNT, REGISTER_COUNT, FIRST_REGISTER) \
+    /* Popularize all the registers with random data */                      \
+    INDEXED_FOR_EACH(FILL_REGISTER, REGISTER_LIST)                           \
+    /* Call the function */                                                  \
+    CALL_A_FUNCTION(FUNCTION_NAME)                                           \
+    /* Restore the stack */                                                  \
+    RESTORE_STACK(STACK_VALUE_COUNT)                                         \
+    /* Restore all the registers */                                          \
+    REVERSE_FOR_EACH(POP_REGISTER, REGISTER_LIST)                            \
+    :: "X"(Generated)                                                        \
+    : COMMA_SEPARATED_FOR_EACH(STRINGIFY, REGISTER_LIST), "memory"           \
+  );                                                                         \
+                                                                             \
+  free(Generated);                                                           \
+}
```

Comment:
This does not need to be a macro, you can pass in a `void *` as argument and call that.

Reply:
It's A LOT simpler to "call" a compile-time function from the inline assembly over the runtime pointer. At least as far as my search went.

I didn't want to imitate templates and spawn a bunch of similar function for each test type as well, but the other option seems even worse.

Reply:
Ah because you don't want to waste a register, uh?
In CISC architecture we can probably get away with something, but in RISC.
Gosh... This is tricky to get right.

Please add a comment stating why this needs to be a distinct function.

Reply:
Believe me, I would have never chosen to ~~suffer~~ create macro hell if I saw another viable alternative.

Reply:
Ok, I have an idea, maybe we can avoid this extra macro by using a global variable-trick or something.

The question is which evil do you hate more? xD

Reply:
The issue is that in RISC architectures you call a function pointer in a global variable without materializing it in a register.
Either we define a *scratch* register for each architecture, or we need to stick with something like this, I'm afraid.

Reply:
That's what I was afraid of, I guess this thread can be closed then.